SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialHandoverManager.cpp

```diff
@@ -0,0 +1,167 @@
+{
+}
+
+void FSpatialHandoverManager::Advance()
+{
+	for (const EntityDelta& Delta : PartitionView.GetViewDelta().EntityDeltas)
+	{
+		switch (Delta.Type)
+		{
+		case EntityDelta::ADD:
+			PartitionsToACK.Add(Delta.EntityId);
+			OwnedPartitions.Add(Delta.EntityId);
+			break;
+		case EntityDelta::REMOVE:
+			OwnedPartitions.Remove(Delta.EntityId);
+			break;
+		case EntityDelta::TEMPORARILY_REMOVED:
+
+			break;
+		default:
+			break;
+		}
+	}
+
+	for (const EntityDelta& Delta : ActorView.GetViewDelta().EntityDeltas)
+	{
+		switch (Delta.Type)
+		{
+		case EntityDelta::UPDATE:
+		{
+			for (const ComponentChange& Change : Delta.ComponentUpdates)
+			{
+				ApplyComponentUpdate(Delta.EntityId, Change.ComponentId, Change.Update);
+			}
+			for (const ComponentChange& Change : Delta.ComponentsRefreshed)
+			{
+				ApplyComponentRefresh(Delta.EntityId, Change.ComponentId, Change.CompleteUpdate.Data);
+			}
+			break;
+		}
+		case EntityDelta::ADD:
+			PopulateDataStore(Delta.EntityId);
+			break;
+		case EntityDelta::REMOVE:
+			DataStore.Remove(Delta.EntityId);
+			break;
+		case EntityDelta::TEMPORARILY_REMOVED:
+			DataStore.Remove(Delta.EntityId);
+			PopulateDataStore(Delta.EntityId);
+			break;
+		default:
+			break;
+		}
+	}
+}
+
+void FSpatialHandoverManager::PopulateDataStore(const Worker_EntityId EntityId)
+{
+	LBComponents2& Components = DataStore.Emplace(EntityId, LBComponents2{});
+	for (const ComponentData& Data : ActorView.GetView()[EntityId].Components)
+	{
+		switch (Data.GetComponentId())
+		{
+		case SpatialConstants::AUTHORITY_INTENTV2_COMPONENT_ID:
+			Components.Intent = AuthorityIntentV2(Data.GetUnderlying());
+			break;
+		case SpatialConstants::AUTHORITY_INTENT_ACK_COMPONENT_ID:
+			Components.IntentACK = AuthorityIntentACK(Data.GetUnderlying());
+			break;
+		default:
+			break;
+		}
+	}
+	HandleChange(EntityId, Components);
+}
+
+void FSpatialHandoverManager::HandleChange(Worker_EntityId EntityId, const LBComponents2& Components)
+{
+	if (Components.Intent.AssignmentCounter != Components.IntentACK.AssignmentCounter)
+	{
+		if (OwnedPartitions.Contains(Components.Intent.PartitionId))
+		{
+			ActorsToACK.Add(EntityId);
+		}
+		else
+		{
+			ActorsToHandover.Add(EntityId);
+		}
+	}
+}
+
+void FSpatialHandoverManager::ApplyComponentUpdate(const Worker_EntityId EntityId, const Worker_ComponentId ComponentId,
+												   Schema_ComponentUpdate* Update)
+{
+	LBComponents2& Components = DataStore[EntityId];
+	switch (ComponentId)
+	{
+	case SpatialConstants::AUTHORITY_INTENTV2_COMPONENT_ID:
+		Components.Intent.ApplyComponentUpdate(Update);
+		break;
+	default:
+		break;
+	}
+
+	HandleChange(EntityId, Components);
+}
+
+void FSpatialHandoverManager::ApplyComponentRefresh(const Worker_EntityId EntityId, const Worker_ComponentId ComponentId,
+													Schema_ComponentData* Data)
+{
+	LBComponents2& Components = DataStore[EntityId];
+	switch (ComponentId)
+	{
+	case SpatialConstants::AUTHORITY_INTENTV2_COMPONENT_ID:
```

Comment:
Likewise we assume this is never possible because the server should never lose interest in an entity it is auth on. Should we warn as a precaution?

Reply:
If it's broken it should always be an error but I'm fairly sure that we have to assume that completeness works.

Reply:
My point being, if this functionality ran, it would suggest somethings gone wrong. When writing complicated distributed systems more logs are better than few.